testing-overhead/Dockerfile-petclinic-base

```diff
@@ -0,0 +1,29 @@
+FROM adoptopenjdk:11-jdk
+
+# This is the base image that will contain a built version of the spring-petclinic-rest
+# application. Installing the dependencies and maven compiling the application is time
+# consuming, so we do it in a base image to save time nightly.
+
+# Allow the test runner to ssh into our container to start and control
+# the application under test.
+
+RUN apt update && \
+    apt install -y openssh-server sudo && \
```

Comment:
Why do we need ssh here?

Reply:
ssh will be used to start/stop the application under test (petclinic-rest) and also to run `jcmd` on the same container.

Reply:
- Usual "Spring Boot in a Docker" package starts application right away.
- You don't need `jcmd` just to get a jfr recording
- `docker-java` probably allows for command execution inside container in a cleaner way that ssh from java :)

Reply:
* Each test pass wants to run the application twice -- typically with agent and without agent (or possibly with two different agent versions). The test runner needs a mechanism to be able to shut down the running application so that it can restart it in the other mode within the same contianer. SSH seemed like a reasonable way to do this, but I'm open to other ideas. I don't think that "just let docker run the jvm for you" is a good answer here.
* I had also expected to use `jcmd` to take a heap dump of the running application before shutting it down.
* Hmmm...so you think we should create yet another java project that requires compile/build and a runtime JDK just to talk to docker in order to run commands? I don't think we want to be communicating with the docker daemon from CI/docker.

SSH is also a convenient way for pulling output files out of the target app's container...but if we hate ssh now I can look into other options for remote command execution and file transfer. A shared docker-compose volume might work for the latter.

Reply:
> can restart it in the other mode within the same contianer

Why is it important to have the same container?

Reply:
In a cloud-based CI environment we have no guarantees about where a given container will be running. If we don't run the test app in the same container instance, then we run the risk of spinning up on different hardware, on a different network segment, with different computational neighbors. By reusing the same container we reduce the thrash and improve the comparability of the two runs.

Reply:
Wait, how do you plan to run that container? I assumed so far that we run some GHA job, which runs some script or java application which starts containers as much as needed. In this case everything happens on the same VM, does not it?

Reply:
Your understanding is correct. The end goal is to just run `docker-compose up --exit-code-from testrunner` from a github action. 

There will be 3 containers -- collector, application, and testrunner. The testrunner will start/stop the application runs inside the application container via ssh (or another mechanism if that's rejected). The same application container instance will be (re)used for multiple app runs.

Reply:
But that means we don't need to use one docker container for all runs?

Reply:
Why? I thought I explained above why I think it's important/useful to re-use the same container for successive runs of the application.

Reply:
Or maybe you're trying to say that a docker-compose environment started by a github action will run all containers on the same VM and therefore we don't need to reuse the container?  If so, how do you know that? Is that documented somewhere?

Reply:
Docker compose does not distribute containers across machines. That is what docker swarm for.

If you assume that containers can spawn on different hosts, then how can you assume that ssh from the coordinator script into a container will work?

Reply:
IIUC, the brunt of the logic is implemented in shell, like this one?

https://github.com/breedx-splk/iguanodon/blob/main/bin/run-remote-sequence.sh

Can we use testcontainers instead? We use it for a lot of things, and it makes it pretty easy to run commands in container, copy files, etc I think.

Reply:
Yeah, the thought was that shell stuff might be more likely to benefit other language agents, but I have been convinced to continue using testcontainers for this. Makes sense.

Reply:
You probably still can test other language agents with testcontainers - only the testrunner has to use JVM

Reply:
Yup, this is a great point too. 🤘🏻